src/Engine/ProtoCore/Lang/Type.cs

```diff
@@ -80,6 +91,16 @@ public string ToShortString()
             }
         }
 
+        /// <summary>
+        /// Get the name of Type from the serialized short string
+        /// </summary>
+        /// <param name="displayTypeName">Serialized short string representation of ProtoCore.Type</param>
+        /// <returns></returns>
+        public static string FromShortString(string displayTypeName)
+        {
+            return displayTypeName.Split('[')[0];
```

Comment:
under what input does this work? This doesn't seem correct.

Reply:
@mjkkirschner Refer to ToShortString(), it seems the actual serialized DisplayTypeName is alway the actual `Name` + RankString which always start with [].

Reply:
@QilongTang that is not true.
see https://github.com/DynamoDS/Dynamo/blob/60fbff3ff4628f4d696ac054b9beea0655282869/src/Engine/ProtoCore/Lang/Type.cs#L38

only if the type is indexable does it have a "[" which makes sense. If the type is a number or other object type - it is not indexable - 

for example - if you build an input like `x:double = 3` I don't think the type will have a `[]` it will just be `double`.

Reply:
hmm.. Yes, but in that `double` example, "double".split('[') still returns "double" right? So at least the Name is deserialized correctly, are you indicating we are missing the `IsIndexable` or `Rank` in schema for more complex case?

Reply:
I am not clear what the output should be for more complex types, is the type name for double[] actually double? or should it be the full type name?

Why did you implement chopping off the rank in the first place?

Reply:
This method is for the ProtoCore.Type struct constructor, because the deserialized could not convert json string back to ProtoCore.Type directly. So we need to convert the `DisplayTypeName` back to an instance of `ProtoCore.Type` somehow without knowing if it is indexable or not. The `DisplayTypeName` itself is not necessarily the true `Name` of Type like you indicated, so I am always chopping off the rank string if there is any. But `IsIndexable` will probably also based on if there is "[]", not sure if this covers all the case but currently DynamoCore need to guess out the original Proto.Type based on the `DisplayTypeName`

Reply:
with this level of complexity why not just also serialize the whole input string and use that to create the parameter?

Reply:
@ke-yu and I talked about this data on slack last week, and we decided we should just serialize the type and rank separately, and let the UI create the display string.

Greg Marr
I'm wondering if we can flatten this, and add the necessary things to the schema.  Perhaps have just `TypeName` and `Rank`, and generate `DisplayTypeName` as needed.
I don't like having things in the Graph part that aren't in the schema.

Yu Ke
Sure thing.  Agree, and these can by generated at runtime.

Reply:
@gregmarr That's great news. If we can serialize TypeName and RankString separately, implementation would be much cleaner particularly about how to assemble back the ProtoCore.Type. Let me know when you guys are done with the schema changes, I will update the Dynamo Part in this PR.

Reply:
How about this

    Parameter:
     properties:
      Name:
        type: string
        description: The name of the parameter.
      TypeName:
        type: string
        description: The name of the parameter type.
      TypeRank:
        type: int
        description: The rank of the parameter type.
      DefaultValue:
        type: string
        description: The default value if the parameter is not provided.

Reply:
@gregmarr @ke-yu @mjkkirschner What do you think the following? This way we do not need to introduce converter and also relatively easy to assemble back a ProtoCore.Type. The proposed schema in the comment above will probably introduce a converter dedicated for Type so that `TypeName` and `TypeRank` can be serialized at the same level.
```json
      "Parameter": {
        "Name": "x",
        "Type": {
          "Name": "var",
          "rank": -1,
        },
        "DefaultValue": "Autodesk.DesignScript.Geometry.Point.ByCoordinates(0, 0, 0)"
      }
```

Reply:
the converter is simple for the above schema I think, but I think @QilongTang is just trying to avoid needing one.

Reply:
@gregmarr @mjkkirschner The converter is simple but @ikeough mentioned earlier that we should try to avoid making new converters and use just JsonAttributes as much as we can. That's the only reason why I am a little bit hesitate to introduce a new one here. If you guys are fine with it, I'll make it

Reply:
My personal preference is to not introduce another type just to match the object nesting of Dynamo.

In fact, I'd prefer to collapse all of `Parameter` into `InputNode`:

    InputNode:
     description: An input to a nested `Graph`.
     allOf:
     - $ref: "#/definitions/Node"
     - required:
      - Name
      - TypeName
      properties:
       Name:
        type: string
        description: The name of the parameter.
       TypeName:
        type: string
        description: The name of the parameter type.
       TypeRank:
        type: integer
        description: The rank of the parameter type.
       DefaultValue:
        type: string
        description: The default value if the parameter is not provided.

 unless it's going to be used somewhere else, such as `OutputNode`

    InputNode:
     allOf:
     - $ref: "#/definitions/Node"
     - required:
      - Parameter
      properties:
        Parameter:
          $ref: "#/definitions/Parameter"

    OutputNode:
     allOf:
     - $ref: "#/definitions/Node"
     - required:
       - Symbol
       properties:
         Symbol:
          $ref: "#/definitions/Parameter"

@ke-yu Were you intending to reuse `Parameter` in this way?